### ISSUE TITLE
Support multiple 'default' groups

### ISSUE THREAD

**Issue by backjo** [keikoproj/instance-manager#189]
**Is this a BUG REPORT or FEATURE REQUEST?**:
Feature Request
**What happened**:
As discussed in #184 and other PRs, it would be nice to be able to leverage configuration boundaries with multiple 'default' options. Use cases:
- Wanting to restrict the `image` being used, but allowing individual IGs to specify which default one they want. I.E. IGs should be able to specify that they whether they want Windows or Linux nodes, but should not be able to choose which AMI is used for either platform.

**Comment by eytan-avisror** [comment]
Option:
Add following spec ->
```yaml
spec:
  eks:
    configuration:

      # used by default
      image: ami-1234567890

      # if not nil, overrides image field
      images:
      - family: windows
        image: ami-11111111111
      - family: amazonlinux2
        image: ami-222222222
      - family: bottlerocket
        image: ami-33333333
```

We can then do matching with the annotation you added in last PRs, so that it allows choosing (default amazonlinux2).
You could then use boundaries in order to make `images` a restricted field, and define these AMIs in the config map.

I think the only downside with this design is that it's only really usable for when you use the boundaries.

Suggestions?

**Comment by backjo** [comment]
That seems pretty clean - and would solve what I'm trying to accomplish. Let me think on it for a bit  - I wonder if there are higher level use cases where folks might want to have different defaults for SGs, subnets or other properties.

**Comment by backjo** [comment]
Another potential source of variability could be OS architecture - i.e. having a separate default image defined for ARM vs x86_64

**Comment by eytan-avisror** [comment]
Could you provide an example use-case?
Would you like to choose an AMI based on both OS and architecture?
For example, bottlerocket on AMD64?

**Comment by backjo** [comment]
Yep, that's one use case. Another is for Windows images, where there are 3 or 4 versions of the OS (two or three support channels, and then Core vs Full) that could be selected (and the windows containers can only run on a subset, depending on their version and the tools that they use).

**Comment by eytan-avisror** [comment]
This seems like a lot of options.. feel free to suggest a design if you have one in mind..

**Comment by backjo** [comment]
How about a new `InstanceGroupTemplate` CR
```yaml
kind: InstanceGroupTemplate
metadata:
 name: my-ig-template
 annotations:
   instancemgr.keikoproj.io/os-family: "windows"
spec:
  ig:
    eks:
      configuration:
        labels:
          test: bdd-test-rolling
        taints:
        - key: node-role.kubernetes.io/bdd-test
          value: bdd-test
          effect: NoSchedule
        clusterName: {{ .ClusterName }}
        subnets: {{range $element := .Subnets}}
          - {{$element}}
        {{ end }}
        keyPairName: {{ .KeyPairName }}
        image: {{ .AmiID }}
        instanceType: t2.small
        volumes:
        - name: /dev/xvdb
          type: gp2
          size: 30
          mountOptions:
            fileSystem: xfs
            mount: /data
        securityGroups: {{range $element := .NodeSecurityGroups}}
          - {{$element}}
        {{ end }}
        tags:
        - key: k8s-minion-manager
          value: use-spot
        metricsCollection:
        - all
  boundary:
    shared:
      mergeOverride:
      - spec.eks.configuration.tags
      merge:
      - spec.strategy
      - spec.eks.configuration.securityGroups
      - spec.eks.configuration.labels
      - spec.eks.configuration.volumes
    restricted:
    - spec.provisioner
    - spec.eks.configuration.clusterName
    - spec.eks.configuration.image
    - spec.eks.configuration.roleName
    - spec.eks.configuration.instanceProfileName
    - spec.eks.configuration.keyPairName
    - spec.eks.configuration.subnets
    - spec.eks.configuration.taints
```

Then, an IG could be specified similar to today, but with a 'template' field to denote which template to use. 
```yaml
kind: InstanceGroup
metadata:
  name: my-ig
spec:
  template: my-ig-template
  eks:
    minSize: 1
    maxSize: 4
```

Advantages of this approach:
- Multiple boundaries can now be defined. Since the boundary is defined per-template, boundaries can be stricter/more relaxed per template, depending on the use case. This would be useful for cluster operators who might want to have harder multi-tenancy, where they could different restricted SGs or Subnets per-template. 
- Access to templates/defaults can be enforced using OPA or other tools - so as a cluster operator, I could choose to expose 'windows" nodes only to certain namespaces based on label.
- Flexibility. Templates can be as coarse or granular as desired.

**Comment by eytan-avisror** [comment]
@backjo This seems like significant change to existing functionality which might be difficult to keep backwards compatible.

The suggestion is interesting (I like the concept of a template object), but need a better understanding of how this will be backwards compatible - the current implementation defines both boundaries and defaults in a configmap, are you suggesting that `InstanceGroupTemplate` objects have both a template and boundaries, where the configmap change to only holding the defaults? or are you suggesting replacing the configmap implementation with the template objects?

Wonder if there might be a simpler approach for the time being that can answer this use case.

Perhaps an approach that is more easily compatible and requires less change, can be adding support for having multiple configmaps with their own defaults and boundaries, and possibly allow the CR to chose the 'template' that way.

WDYT?

**Comment by eytan-avisror** [comment]
Gave this some more thought.. here is another possibility on how to have this capability

Our configmap currently looks something like this:
```yaml
data:
  defaults: |
    spec:
      eks:
        configuration:
        # used by default
          image: ami-1234567890
  boundaries: |
    < .. boundary config .. >

# we can introduce a third option called "conditionals" which can rely on annotations/labels
  conditionals: 
  - annotation:  "instancemgr.keikoproj.io/os-family=windows"
    defaults: | 
      spec:
        eks:
          configuration:
            # a windows image
            image: ami-22222222222
```

Users can also customize this by adding their own annotations to override defaults if they wish e.g. 
```yaml
  conditionals: 
  - annotation:  "custom-network-zone=zone-a"
    defaults: | 
      spec:
        eks:
          configuration:
            subnets:
            - subnet-1234567
            securityGroups:
            - sg-1234567890
```


Conditional default values will take precedence over standard default, in the case of merge/replace boundaries, custom resource fields are always first priority, conditionals will take priority between custom resource values and defaults values

This would be a generic implementation that will allow us to match a default configuration with an annotation or label, and would be relatively easy to add to existing implementation.

**Comment by backjo** [comment]
@eytan-avisror I like it - being able to compose configuration snippets provides a lot of flexibility - it should allow us to get rid of a bunch of copy/pasted settings that we have right now.

**Comment by backjo** [comment]
@eytan-avisror are you working on this? Have some bandwidth to try it out, but don't want to step on your feet if you're already on it :)

**Comment by eytan-avisror** [comment]
@backjo go for it! I haven't started working on this yet.
This is probably a good starting point: https://github.com/keikoproj/instance-manager/blob/master/controllers/provisioners/config.go#L109-L129

**Comment by eytan-avisror** [comment]
closed by #273
